reader.go

```diff
@@ -55,67 +100,200 @@ const (
 
-	return errG.Wait()
+			if err := r.config.ProgressManager.StartGeneration(gen.startTime); err != nil {
+				return err
+			}
+
+			// Start batch readers for this generation
+			split, err := r.splitStreams(gen.streams)
+			if err != nil {
+				return err
+			}
+
+			l.Printf("grouped %d streams into %d batches", len(gen.streams), len(split))
+
+			genErrG, genCtx := errgroup.WithContext(runCtx)
+
+			readers := make([]*streamBatchReader, 0, len(split)*len(r.config.TableNames))
+			for _, tableName := range r.config.TableNames {
+				// TODO: This is ugly?
+				splitName := strings.SplitN(tableName, ".", 2)
+				for _, group := range split {
+					readers = append(readers, newStreamBatchReader(
+						r.config,
+						gen.startTime,
+						group,
+						splitName[0],
+						splitName[1],
+						gocql.MinTimeUUID(r.readFrom),
+					))
+				}
+			}
+
+			sleepAmount := r.config.Advanced.PostNonEmptyQueryDelay / time.Duration(len(readers))
+			for i := range readers {
+				reader := readers[i] // TODO: Should this be interruptible?
+				<-time.After(sleepAmount)
+				genErrG.Go(func() error {
+					return reader.run(genCtx)
+				})
+			}
+
+			var nextGen *generation
+			genErrG.Go(func() error {
+				var err error
+				nextGen, err = r.genFetcher.Get(genCtx)
+				if err != nil {
+					return err
+				}
+				for _, reader := range readers {
+					if nextGen == nil {
+						// The reader was stopped
+						stopAt, _ := r.stopTime.Load().(time.Time)
+						if stopAt.IsZero() {
+							reader.stopNow()
+						} else {
+							reader.close(gocql.MaxTimeUUID(stopAt))
+							r.readFrom = stopAt
+						}
+					} else {
+						reader.close(gocql.MinTimeUUID(nextGen.startTime))
+						r.readFrom = nextGen.startTime
+					}
+				}
+				return nil
+			})
+
+			if err := genErrG.Wait(); err != nil {
+				return err
+			}
+			l.Printf("stopped reading from generation %v", gen.startTime)
+			if nextGen == nil {
+				break
+			}
+			gen = nextGen
+
+			if err := r.config.ProgressManager.EndGeneration(gen.startTime); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	return runErrG.Wait()
 }
 
-// Stop gracefully stops the CDC reader. It does not wait until the reader shuts down.
+// Stop tells the reader to stop as soon as possible. There is no guarantee
+// related to how much data will be processed in each stream when the reader
+// stops. If you want to e.g. make sure that all cdc log data with timestamps
+// up to the current moment was processed, use (*Reader).StopAt(time.Now()).
+// This function does not wait until the reader stops.
 func (r *Reader) Stop() {
-	atomic.StoreInt32(&r.stopped, 1)
+	close(r.stoppedCh)
+}
+
+// StopAt tells the reader to stop reading changes after reaching given timestamp.
+// Does not guarantee that the reader won't read any changes after the timestamp,
+// but the reader will stop after all tables and streams are advanced to or past
+// the timestamp.
+// This function does not wait until the reader stops.
```

Comment:
And if one wants to wait, how does he do that?

Reply:
They would wait until the call to `(*Reader).Run(...)` finishes.